imageio/core/imopen.py

```diff
@@ -100,7 +101,15 @@ def imopen(
 
     plugin : str, Plugin, or None
         The plugin to use. If set to None (default) imopen will perform a
-        search for a matching plugin.
+        search for a matching plugin. If not None, this takes priority over
+        the provided format hint.
+    format_hint : str
+        A format hint to help optimize plugin selection given as the format's
+        extension, e.g. ".png". This can speed up the selection process for
+        ImageResources that don't have an explicit extension, e.g. streams, or
+        for ImageResources where the extension does not match the resource's
+        content. If the ImageResource lacks an explicit extension, it will be
+        set to this format.
```

Comment:
What about if `plugin` starts with a dot, it means an extension, and thus a format hint? This is how it worked previously.

Reply:
> What about if plugin starts with a dot, it means an extension, and thus a format hint?

Yes. I separated them, because plugins and formats are different things. This is ultimately how we ended up with hybrid format names like `PNG-FI`, `PNG-PIL`, but also with format names (that are actually plugins) `FFMPEG` or `SPE`. Separating them into `plugin=` (only plugin names allowed here) and `format=` (only formats allowed here) makes this cleaner. 

> This is how it worked previously.

Indeed, this is something we introduced to support the v2 API through imopen. It won't work in the v3 API, where the value of `plugins` is checked against ... well plugins (`plugin in config.known_plugins`). If we do end up merging this, I will refactor the internals of the v2 API to separate plugins and formats and send them to the appropriate arguments. This will keep the behavior local to the v2 API and allow us to have a cleaner backend.

What's more is that calling `PNG-PIL` doesn't actually guarantee that the data is treated as a PNG. The v3 pillow plugin currently has no notion of format (it hands a file-like object to pillow and, if that fails, concludes that pillow can't read it). The v2 plugin does have a notion of format (enforced via file extension), but if that fails (e.g. for bytes that have no extension) it also just does the same thing as the v3 plugin: call the backend and see if things fail. Notably, neither plugin restricts the formats pillow may use internally (this could be done though), so at the moment iio.imread(some-jpg-bytes, format="PNG-PIL") will look like it loads a PNG from a JPEG, but it actually just uses pillows JPEG decoder and ignores the `format=` part silently. Because of this, the format is a _hint_ and not enforced. 

I felt like this should be documented as such, in an effort to be more transparent and, while at it, I figured it would be useful to have a similar functionality in v3.

Brownie points: we no longer have to figure out which format supports which extension and create strange virtual formats like `PNG-PIL` and alike. `plugins` is everything listed in `config.known_plugins` and `format_hint` is everything listed in `config.known_extensions` and the rest is mix and match. If the format being hinted is actually not what is in the file, that's totally fine; it's a hint anyway, and all it does is make the selection process slower, because we try unsuitable plugins first due to the wrong hint.

Reply:
Good point. I was a bit reluctant about making the API more complex by introducing a new argument. My initial reaction was that we can still use a single argument, that is a format hint if it's a string starting with a dot, and a plugin selection otherwise, but indeed this becomes a bit weird, since the former is a hint and the latter a hard selection.